pandas/_libs/tslibs/parsing.pyx

```diff
@@ -44,13 +50,73 @@ class DateParseError(ValueError):
 
 _DEFAULT_DATETIME = datetime(1, 1, 1).replace(hour=0, minute=0,
                                               second=0, microsecond=0)
+_DEFAULT_TZINFO = _DEFAULT_DATETIME.tzinfo
 
 cdef:
     object _TIMEPAT = re.compile(r'^([01]?[0-9]|2[0-3]):([0-5][0-9])')
 
     set _not_datelike_strings = {'a', 'A', 'm', 'M', 'p', 'P', 't', 'T'}
 
 # ----------------------------------------------------------------------
+DEF delimiters = b' /-\\'
+DEF MAX_DAYS_IN_MONTH = 31
+DEF MAX_MONTH = 12
+
+cdef bint _is_not_delimiter(const char ch):
+    return strchr(delimiters, ch) == NULL
+
+cdef inline int _parse_2digit(const char* s):
+    cdef int result = 0
+    result += getdigit_ascii(s[0], -10) * 10
+    result += getdigit_ascii(s[1], -100) * 1
+    return result
+
+cdef inline int _parse_4digit(const char* s):
```

Comment:
is this really more performant than `int(s)`?  It looks like this is only used if we already know `s` is 4 characters.  I guess we don't know that it is all digits?

Reply:
If you want microbenchmarks I can try to make those, or you can just take my word on it - `int(s)` is _very_ slow for non-integer strings because it uses Python mechanism of raising and catching exceptions, which itself costs at least 10x more than parsing the integer (because Python has to allocate memory, format error message and walk a list of frames to construct the traceback - everything which would be just thrown away here, because the only fact that we care here is "if it isn't an integer - return `None`").

Reply:
So a quick measurement shows this (figures are from the benchmark presented in this PR):

sample date| _parse_4digit | int| parse/int ratio
-|-|-|-
03/2010|45.4 ms|54.6 ms|0.83
01/02/2019|17.4 ms|21.8 ms|0.80

Sample implementation:
```cython
cdef inline object parse_delimited_date(object date_string, bint dayfirst,
                                        object tzinfo):
    cdef:
        int day = 1, month = 1, year
        int length = len(date_string)

    if length == 10:
        if date_string[2] not in ' /-\\' or date_string[5] not in ' /-\\':
            return None, None
        try:
            day = int(date_string[:2], 10)
            month = int(date_string[3:5], 10)
            year = int(date_string[6:], 10)
        except ValueError:
            return None, None
        reso = 'day'
    elif length == 7:
        if date_string[2] not in ' /-\\':
            return None, None
        try:
            month = int(date_string[:2], 10)
            year = int(date_string[3:], 10)
        except ValueError:
            return None, None
        reso = 'month'
    else:
        return None, None

    if 1 <= month <= MAX_DAYS_IN_MONTH and 1 <= day <= MAX_DAYS_IN_MONTH \
            and (month <= MAX_MONTH or day <= MAX_MONTH):
        if month > MAX_MONTH or (day < MAX_MONTH and dayfirst):
            day, month = month, day
        return PyDateTimeAPI.DateTime_FromDateAndTime(year, month, day,
                0, 0, 0, 0, tzinfo, PyDateTimeAPI.DateTimeType), reso

    raise DateParseError("Invalid date specified (%d/%d)" %
                            (month, day))
```

I won't say it's much easier, but it's at least 20% slower.